funsor/cnf.py

```diff
@@ -231,14 +231,14 @@ def eager_contraction_to_binary(red_op, bin_op, reduced_vars, lhs, rhs):
 def eager_contraction_tensor(red_op, bin_op, reduced_vars, *terms):
     if not all(term.dtype == "real" for term in terms):
         raise NotImplementedError('TODO')
-    return _eager_contract_tensors(reduced_vars, terms, backend="torch")
+    return _eager_contract_tensors(reduced_vars, terms, backend=terms[0].backend)
```

Comment:
`Tensor.backend` is a string of `"torch"` or `"numpy"`.

Reply:
I think it is better to put this in a new file funsor/einsum/numpy_log.py. Note the `opt_einsum` requirements are a little subtle: they use a `tensordot`, an `einsum` and a `transpose` implementation. Since you've already implemented a logsumexp `einsum`, you can create a `tensordot` via
```py
from pyro.ops.einsum.util import Tensordot
tensordot = Tensordot(einsum)
```